README.md
# bump_version

This is a tool for bumping version numbers in Go files.

## Installation

For the moment, you'll need a working Go installation.

```
go get github.com/kevinburke/bump_version
```

That will install the `bump_version` binary to your `$GOPATH`.

## Usage

```
bump_version <major|minor|patch> <filename>
```

This will:

1. Look for a `const` named `version`, `VERSION`, or `Version` in that file.
   Here's an example:

    ```go
    package main

    const VERSION = "0.2.1"
    ```

2. Apply the version bump - `bump_version major` will increment the major
version number, `bump_version minor` will increment the middle version number,
`bump_version patch` will increment the last version number. If your version is
"0.3" and you ask for `bump_version minor`, the new version will be "0.4".

3. Write the new file to disk, with the bumped version.

4. Add the file with `git add <filename>`.

5. Add a commit with the message "x.y.z" (`git commit -m "<new_version>"`)

6. Tag the new version.

If any of these steps fail, `bump_version` will abort.

## Notes

The VERSION should be a string in one of these formats: "3", "0.3",
"0.3.4". Any prefixes like "v" or suffixes like "0.3.3-beta" will be
stripped or generate an error.

- `"v0.1"` - parse error, no prefixes allowed.
- `bump_version("0.1", "minor")` -> "0.2"
- `bump_version("0.1", "patch")` -> "0.1.1"
- `bump_version("0.1", "major")` -> "1.1"
- `bump_version("0.1-beta", "major")` -> "1.1"
- `bump_version("devel", "major")` -> parse error.

We use the VERSION in code exclusively - any existing git tags are ignored.

Alan Shreve would like to note that you probably shouldn't store version
numbers in code - instead, check in `const VERSION = "devel"`, then build your
project via:

```
go build -ldflags="-X main.VERSION=0.2"
```

Which you are welcome to do!


CHANGELOG.md
# Version 2.2

We now support parsing a Version constant in either a `const` or a `var`
declaration.


lib/lib_test.go
package bump_version

import (
	"reflect"
	"testing"
)

func TestChangeVersion(t *testing.T) {
	testCases := []struct {
		in    string
		vtype VersionType
		out   string
	}{
		{"0.4", Major, "1.0"},
		{"0.4.0", Major, "1.0.0"},
		{"1.0", Major, "2.0"},
		{"1", Major, "2"},
		{"1.0.1", Minor, "1.1.0"},
	}
	for _, tt := range testCases {
		v, err := changeVersion(tt.vtype, tt.in)
		if err != nil {
			t.Fatal(err)
		}
		if v.String() != tt.out {
			t.Errorf("changeVersion(%s, %s): got %s, want %s", tt.vtype, tt.in, v.String(), tt.out)
		}
	}
}

func TestVersionString(t *testing.T) {
	typ := reflect.TypeOf(VERSION)
	if typ.String() != "string" {
		t.Errorf("expected VERSION to be a string, got %#v (type %#v)", VERSION, typ.String())
	}
}

func TestString(t *testing.T) {
	v := Version{0, 0, 1}
	if v.String() != "0.0.1" {
		t.Errorf("wrong version string reported")
	}
	v = Version{1, 0, 1}
	if v.String() != "1.0.1" {
		t.Errorf("wrong version string reported")
	}
	v = Version{1, -1, -1}
	if want := "1"; v.String() != want {
		t.Errorf("wrong version string reported: got %q want %q", v.String(), want)
	}
}

var lessTests = []struct {
	i, j string
	want bool
}{
	{"1", "2", true},
	{"1.1", "2", true},
	{"1.3.7", "2", true},
	{"1.3.7", "0.1", false},
	{"1.3.7", "1.3.8", true},
	{"1.3.7", "1.3.6", false},
	{"1.3.7", "1.3.7", false},
	{"1.3.7", "1", false},
}

func TestLess(t *testing.T) {
	for _, tt := range lessTests {
		i, _ := Parse(tt.i)
		j, _ := Parse(tt.j)
		got := Less(i, j)
		if got != tt.want {
			t.Errorf("Less(%q, %q): got %t, want %t", i, j, got, tt.want)
		}
	}
}


current_version/main.go
// The current_version binary prints out the value of a version number stored in
// a Go source file.
package main

import (
	"flag"
	"fmt"
	"os"

	bump_version "github.com/kevinburke/bump_version/lib"
)

func usage() {
	fmt.Fprintf(os.Stderr, "usage: current_version <filename>\n")
	flag.PrintDefaults()
}

func _main(flags *flag.FlagSet, cmdArgs []string) int {
	if err := flags.Parse(cmdArgs); err != nil {
		flags.Usage()
		return 2
	}
	args := flags.Args()
	if len(args) != 1 {
		flags.Usage()
		return 2
	}
	filename := args[0]
	version, err := bump_version.GetInFile(filename)
	if err != nil {
		os.Stderr.WriteString(err.Error() + "\n")
		return 2
	}
	os.Stdout.WriteString(version.String() + "\n")
	return 0
}

func main() {
	flag.Usage = usage
	os.Exit(_main(flag.CommandLine, os.Args[1:]))
}


lib/lib.go
// Package bump_version makes it easy to parse and update version numbers in Go
// source files.
package bump_version

import (
	"errors"
	"fmt"
	"go/ast"
	"go/parser"
	"go/printer"
	"go/token"
	"os"
	"strconv"
	"strings"
)

const VERSION = "2.2"

type VersionType string

const (
	Major = VersionType("major")
	Minor = VersionType("minor")
	Patch = VersionType("patch")
)

func ValidVersionType(vtype VersionType) bool {
	switch vtype {
	case Major, Minor, Patch:
		return true
	default:
		return false
	}
}

type Version struct {
	Major int64
	// May be "-1" to signify that this version field is unused.
	Minor int64
	Patch int64
}

func (v Version) String() string {
	if v.Patch >= 0 {
		return fmt.Sprintf("%d.%d.%d", v.Major, v.Minor, v.Patch)
	}
	if v.Minor >= 0 {
		return fmt.Sprintf("%d.%d", v.Major, v.Minor)
	}
	if v.Major >= 0 {
		return fmt.Sprintf("%d", v.Major)
	}
	return "%!s(INVALID_VERSION)"
}

// Parse parses a version string of the forms "2", "2.3", or "0.10.11". Any
// information after the third number ("2.0.0-beta") is discarded. Very little
// effort is taken to validate the input.
//
// If a field is omitted from the string version (e.g. "0.2"), it's stored in
// the Version struct as the integer -1.
func Parse(version string) (Version, error) {
	if len(version) == 0 {
		return Version{}, errors.New("bump_version: empty version string")
	}

	parts := strings.SplitN(version, ".", 3)
	if len(parts) == 1 {
		major, err := strconv.ParseInt(parts[0], 10, 64)
		if err != nil {
			return Version{}, err
		}
		return Version{
			Major: major,
			Minor: -1,
			Patch: -1,
		}, nil
	}
	if len(parts) == 2 {
		major, err := strconv.ParseInt(parts[0], 10, 64)
		if err != nil {
			return Version{}, err
		}
		minor, err := strconv.ParseInt(parts[1], 10, 64)
		if err != nil {
			return Version{}, err
		}
		return Version{
			Major: major,
			Minor: minor,
			Patch: -1,
		}, nil
	}
	if len(parts) == 3 {
		major, err := strconv.ParseInt(parts[0], 10, 64)
		if err != nil {
			return Version{}, err
		}
		minor, err := strconv.ParseInt(parts[1], 10, 64)
		if err != nil {
			return Version{}, err
		}
		patchParts := strings.SplitN(parts[2], "-", 2)
		patch, err := strconv.ParseInt(patchParts[0], 10, 64)
		if err != nil {
			return Version{}, err
		}
		return Version{
			Major: major,
			Minor: minor,
			Patch: patch,
		}, nil
	}
	return Version{}, fmt.Errorf("bump_version: invalid version string: %q", version)
}

// Bump increments the version number by the given vtype (major/minor/patch).
// Bump panics if vtype is not a known VersionType.
func Bump(version Version, vtype VersionType) Version {
	switch vtype {
	case Major:
		version.Major++
		if version.Minor != -1 {
			version.Minor = 0
		}
		if version.Patch != -1 {
			version.Patch = 0
		}
		return version
	case Minor:
		if version.Minor == -1 {
			version.Minor = 0
		}
		if version.Patch != -1 {
			version.Patch = 0
		}
		version.Minor++
		return version
	case Patch:
		if version.Patch == -1 {
			version.Patch = 0
		}
		version.Patch++
		return version
	default:
		panic(fmt.Sprintf("bump_version: invalid version type: %s", vtype))
	}
}

// changeVersion takes a basic literal representing a string version, and
// increments the version number per the given VersionType.
func changeVersion(vtype VersionType, value string) (Version, error) {
	versionNoQuotes := strings.Replace(value, "\"", "", -1)
	version, err := Parse(versionNoQuotes)
	if err != nil {
		return Version{}, err
	}
	return Bump(version, vtype), nil
}

func findBasicLit(file *ast.File) (*ast.BasicLit, error) {
	for _, decl := range file.Decls {
		gd, ok := decl.(*ast.GenDecl)
		if !ok {
			continue
		}
		if gd.Tok != token.CONST && gd.Tok != token.VAR {
			continue
		}
		spec, ok := gd.Specs[0].(*ast.ValueSpec)
		if !ok {
			continue
		}
		if strings.ToUpper(spec.Names[0].Name) != "VERSION" {
			continue
		}
		value, ok := spec.Values[0].(*ast.BasicLit)
		if !ok || value.Kind != token.STRING {
			return nil, fmt.Errorf("bump_version: VERSION constant is not a string, was %#v", value.Value)
		}
		return value, nil
	}
	return nil, errors.New("bump_version: No version const found")
}

func writeFile(filename string, fset *token.FileSet, file *ast.File) error {
	f, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer f.Close()
	cfg := printer.Config{Mode: printer.UseSpaces | printer.TabIndent, Tabwidth: 8}
	return cfg.Fprint(f, fset, file)
}

var errNoChanges = errors.New("bump_version: no changes made")

func changeInFile(filename string, f func(*ast.BasicLit) error) error {
	fset := token.NewFileSet()
	parsedFile, err := parser.ParseFile(fset, filename, nil, parser.ParseComments)
	if err != nil {
		return err
	}
	lit, err := findBasicLit(parsedFile)
	if err != nil {
		return fmt.Errorf("bump_version: no Version const found in %s", filename)
	}
	funcErr := f(lit)
	if funcErr != nil && funcErr != errNoChanges {
		return err
	}
	if funcErr == errNoChanges {
		return nil
	}
	writeErr := writeFile(filename, fset, parsedFile)
	return writeErr
}

// GetInFile returns the version found in the file.
func GetInFile(filename string) (Version, error) {
	var version Version
	err := changeInFile(filename, func(lit *ast.BasicLit) error {
		versionNoQuotes := strings.Replace(lit.Value, "\"", "", -1)
		var err error
		version, err = Parse(versionNoQuotes)
		if err != nil {
			return err
		}
		return errNoChanges
	})
	return version, err
}

// SetInFile sets the version in filename to newVersion.
func SetInFile(newVersion Version, filename string) error {
	return changeInFile(filename, func(lit *ast.BasicLit) error {
		lit.Value = fmt.Sprintf("%q", newVersion.String())
		return nil
	})
}

// BumpInFile finds a constant named VERSION, version, or Version in the file
// with the given filename, increments the version per the given VersionType,
// and writes the file back to disk. Returns the incremented Version object.
func BumpInFile(vtype VersionType, filename string) (Version, error) {
	var version Version
	err := changeInFile(filename, func(lit *ast.BasicLit) error {
		var err error
		version, err = changeVersion(vtype, lit.Value)
		if err != nil {
			return err
		}
		lit.Value = fmt.Sprintf("%q", version.String())
		return nil
	})
	return version, err
}

// Less reports whether i is a lower version number than j.
func Less(i, j Version) bool {
	if i.Major < j.Major {
		return true
	}
	if i.Major > j.Major {
		return false
	}
	if i.Minor < j.Minor {
		return true
	}
	if i.Minor > j.Minor {
		return false
	}
	return i.Patch < j.Patch
}
